src/handover.cpp

```diff
@@ -0,0 +1,220 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "handover.hpp"
+
+using namespace realm;
+
+AnyThreadConfined::AnyThreadConfined(const AnyThreadConfined& thread_confined)
+{
+    switch (thread_confined.m_type) {
+        case Type::Object:
+            new (&m_object) Object(thread_confined.m_object);
+            break;
+
+        case Type::List:
+            new (&m_list) List(thread_confined.m_list);
+            break;
+
+        case Type::Results:
+            new (&m_results) Results(thread_confined.m_results);
+            break;
+    }
+    new (&m_type) Type(thread_confined.m_type);
+}
+
+AnyThreadConfined& AnyThreadConfined::operator=(const AnyThreadConfined& thread_confined)
+{
+    switch (thread_confined.m_type) {
+        case Type::Object:
+            m_object = thread_confined.m_object;
+            break;
+
+        case Type::List:
+            m_list = thread_confined.m_list;
+            break;
+
+        case Type::Results:
+            m_results = thread_confined.m_results;
+            break;
+    }
+    m_type = thread_confined.m_type;
+    return *this;
+}
+
+AnyThreadConfined::AnyThreadConfined(AnyThreadConfined&& thread_confined)
+{
+    switch (thread_confined.m_type) {
+        case Type::Object:
+            new (&m_object) Object(std::move(thread_confined.m_object));
+            break;
+
+        case Type::List:
+            new (&m_list) List(std::move(thread_confined.m_list));
+            break;
+
+        case Type::Results:
+            new (&m_results) Results(std::move(thread_confined.m_results));
+            break;
+    }
+    new (&m_type) Type(std::move(thread_confined.m_type));
+}
+
+AnyThreadConfined& AnyThreadConfined::operator=(AnyThreadConfined&& thread_confined)
+{
+    switch (thread_confined.m_type) {
+        case Type::Object:
+            m_object = std::move(thread_confined.m_object);
+            break;
+
+        case Type::List:
+            m_list = std::move(thread_confined.m_list);
+            break;
+
+        case Type::Results:
+            m_results = std::move(thread_confined.m_results);
+            break;
+    }
+    m_type = std::move(thread_confined.m_type);
+    return *this;
+}
+
+AnyThreadConfined::~AnyThreadConfined()
+{
+    switch (m_type) {
+        case Type::Object:
+            m_object.~Object();
+            break;
+
+        case Type::List:
+            m_list.~List();
+            break;
+
+        case Type::Results:
+            m_results.~Results();
+            break;
+    }
+}
+
+SharedRealm AnyThreadConfined::get_realm() const
+{
+    switch (m_type) {
+        case Type::Object:
+            return m_object.realm();
+
+        case Type::List:
+            return m_list.get_realm();
+
+        case Type::Results:
+            return m_results.get_realm();
+    }
+}
+
+AnyHandover AnyThreadConfined::export_for_handover() const
```

Comment:
There might be potential to reduce copying here a bit. For example, we could `std::move` the `Query` and `Row` objects maybe (if we gave the function move semantics). We might also benefit from using `MutableSourcePayload::Move` since we already have a copy of the `Query` (I assume there's no shared state). I'm not sure how the performance of `std::move` vs. `MutableSourcePayload::Move` would differ though. Thoughts? @tgoyne